#166 chat enhancements

Zoap opened the pull request:
I took the liberty of adding to the fixes in this issue even tho I'm not supposed to (I don't like leaving things broken) and added "Disable player input  when chat is active" as one of the issues to fix in this Issue.

Summary of what has been done:
- Total refactoring of InChatRoom.cs (Resulted in three new classes)
- ChatCommands.cs enum for the chat commands available
- ChatCommandHandler.cs handles everything to do with chat commands and their logic
- ChatUtility.cs static class with methods for formatting chat text

- [x] Implement a scrollbar for chat, and limit the amount of messages that are stored inside the InRoomChat.cs to 100.
- [x] Add a character limit to chat messages.
- [x] Using broken tags should not break the chat.
- [x] Player inputs disabled while chat is focused.

Closes [#166](https://github.com/AoTTG-2/AoTTG-2/issues/166)

Zoap committed:
#166
- set message character limit to 1000
- refactored the entire class InRoomChat
- added new methods to InRoomChat

Zoap committed:
#166 refactoring

Zoap committed:
#166
- Makes Chat scrollable
- Chat message character limit set to 1000
- Chat messages limit set to 100
- New Class for handling chat commands
- General refactoring of InRoomChat.cs

Zoap committed:
#166
- Fixed bug with GetTeam() not casting properly
- Added simple regex for catching bad markup in chat
- Refactoring InRoomChat
- Refactoring ChatCommandHandler

Zoap committed:
all issues resolved

Zoap committed:
removed unused using statments

Zoap committed:
changes regex for markup validation

Zoap commented:
Something I forgot to mention was that the chat is scrollable, but I couldn't for the life of me figure out how to get the scrollbar working, so I never added the scrollbar and was going to pass that to someone else, but as you see I forgot. So... how important is the actual scrollbar?

mwnDK1402 commented on Assets/Scripts/Room/Chat/ChatCommandHandler.cs:
```diff
@@ -0,0 +1,777 @@
+
+    private static void OutputIsRc()
+    {
+        string message;
```
```cs
string message = masterRC ? "Is RC" : "Not RC";
```

mwnDK1402 commented on Assets/Scripts/Room/Chat/ChatCommandHandler.cs:
```diff
@@ -0,0 +1,777 @@
+                        break;
+                    case 1:
+                        SwitchTeam(1);
+                        message = FormatTextColor00FFFF("You have joined team cyan.");
```
Should be able to find a more intuitive way to express this. With an enum like `enum Team { None, Cyan, Magenta }`, for example. Or with a dictionary.

mwnDK1402 commented on Assets/Scripts/Room/Chat/ChatCommandHandler.cs:
```diff
@@ -0,0 +1,777 @@
+
+        ChatCommand command;
+
+        if (Enum.TryParse(commands[0].ToLower(), out command))
```
I think I would prefer a `Dictionary<string, CommandHandler> commandHandlerByName`. Would make this code as simple as this:
```cs
CommandHandler handler;
if (TryGetCommandHandler(command, parameter, out handler) && HasPermissionToUse(handler))
    handler.Execute();
```
... where there would be a `Dictionary<name, CommanHandler> commandHandlerByName` used by `TryGetCommandHandler`.

Zoap replied:
It's a great solution, however, I do believe that we can do that refraction at a later date, maybe when we start adding more commands we could add this as an issue to be resolved, but for administrative purposes I don't think we should resolve it in this branch.

mwnDK1402 commented on Assets/Scripts/Room/Chat/InRoomChat.cs:
```diff
@@ -1,593 +1,229 @@
+    /// <summary>
+    /// Removes messages when exceding max storage
+    /// </summary>
+    private void RemoveMessageIfMoreThenMax()
```
Typo here: `MoreThanMax`.
How about something like `TrimMessages`?

mwnDK1402 commented on Assets/Scripts/Room/Chat/ChatCommandHandler.cs:
```diff
@@ -0,0 +1,777 @@
+                    return;
+                }
+
+                if (OnPrivateServer || isMasterClient)
```
This check seems redundant, since above is an `if (!A)`, and this is checking `if(A)`.

mwnDK1402 commented on Assets/Scripts/Room/Chat/ChatCommandHandler.cs:
```diff
@@ -0,0 +1,777 @@
+    /// </summary>
+    private static void EnterExitSpecMode()
+    {
+        string message;
```
Rename to `ToggleSpecMode()`.
```cs
// This will be more readable when we improve the settings.
settings[0xf5] = ((int)settings[0xf5] == 1 ? 0 : 1);
bool specMode = settings[0xf5] == 1;
instance.EnterSpecMode(specMode);
string message = FormatSystemMessage(specMode ? "You have entered spectator mode." : "You have exited spectator mode.");
instance.chatRoom.AddMessage(message);
```

mwnDK1402 commented on Assets/Scripts/Room/Chat/ChatCommandHandler.cs:
```diff
@@ -0,0 +1,777 @@
+    /// Send private message to player with Id
+    /// </summary>
+    /// <param name="input"></param>
+    private static void SendPrivateMessage(string[] input)
```
This seems super confusing. `string[] input` is just the list of parameters. And we string them all back together, after they've been split up. :P
So `input` should be renamed to `parameters`. `whatIsThisVar` is the actual text pf the message to be sent. Maybe `messageText` or `rawMessage`.
Maybe we should use a `StringBuilder` here.

mwnDK1402 commented on Assets/Scripts/Room/Chat/ChatCommandHandler.cs:
```diff
@@ -0,0 +1,777 @@
+                    int maxPlayers;
+                    float time;
+
+                    switch (command)
```
Same as L662. This one hasn't even been turned into methods, though.

mwnDK1402 commented on Assets/Scripts/Room/Chat/InRoomChat.cs:
```diff
@@ -1,593 +1,229 @@
     public Text ChatText;
 
-    public void addLINE(string newLine)
+    public void OnGUI()
```
Is this the correct Unity Message to use?
https://docs.unity3d.com/ScriptReference/MonoBehaviour.OnGUI.html

Zoap replied:
I have no idea.

mwnDK1402 replied:
```suggestion
    private void Update()
```
> OnGUI is the only function that can implement the "Immediate Mode" GUI (IMGUI) system for rendering and handling GUI events. Your OnGUI implementation might be called several times per frame (one call per event).

mwnDK1402 reviewed (commented):
I tried just looking through it and seeing what things should probably be refactored.

Zoap committed:
resolves comments in PR

Zoap committed:
missed ;

Zoap committed:
Merge from development

Lithrun commented on Assets/Photon Unity Networking/UtilityScripts/PunTeams.cs:
```diff
@@ -102,7 +102,7 @@ public static PunTeams.Team GetTeam(this PhotonPlayer player)
         if (player.CustomProperties.TryGetValue(PunTeams.TeamPlayerProp, out teamId))
         {
-            return (PunTeams.Team)teamId;
+            return (PunTeams.Team)(int)teamId;
```
Is there a reason why this change is introduced? Photon Unity Networking is an unity asset (third party library). We shouldn't modify this.

Zoap replied:
L56 in ChatUtility.cs, the player.GetTeam() method I assumed should return a PunTeams.Team enum, but it didnt until I made the cast in that method, but I see now that I maybe should make the cast outside of method instead. I'll change it.

Zoap replied:
InvalidCastException: Specified cast is not valid.
TeamExtensions.GetTeam (PhotonPlayer player) (at Assets/Photon Unity Networking/UtilityScripts/PunTeams.cs:105)
ChatUtility.SetNameColorDependingOnteam (PhotonPlayer player) (at Assets/Scripts/Room/Chat/ChatUtility.cs:56)
InRoomChat.ChatAll (System.String message) (at Assets/Scripts/Room/Chat/InRoomChat.cs:77)
InRoomChat.HandleChatInput (InRoomChat chat) (at Assets/Scripts/Room/Chat/InRoomChat.cs:116)
InRoomChat.OnGUI () (at Assets/Scripts/Room/Chat/InRoomChat.cs:30)

Zoap replied:
It seems that casting object to PunTeams.Team is not valid for some reason

Lithrun replied:
Let's see. Perhaps the problem is that the PunTeams.Team is of type byte and not of type integer?

Zoap replied:
Yeah but in my switch case I am using Punteams.Team

```
var playerTeam = player.GetTeam();
switch (playerTeam)
{
case PunTeams.Team.red:
    name = FormatTextColor00FFFF(name);
    break;
case PunTeams.Team.blue:
    name = FormatTextColorFF00FF(name);
    break;
default:
    break;
}
```

Lithrun commented on Assets/Scripts/Room/Chat/ChatCommandHandler.cs:
```diff
@@ -0,0 +1,722 @@
+    /// Revive player with Id
+    /// </summary>
+    /// <param name="playerIdString"></param>
+    private static void RevievePlayer(string playerIdString)
```
*Revive

Lithrun commented on Assets/Scripts/Room/Chat/ChatCommandHandler.cs:
```diff
@@ -0,0 +1,722 @@
+
+    private static void ChangeRoomProperties(string[] input)
+    {
+        if (isMasterClient)
```
There's several if statements here. Code can be made more readable with
```
if (!isMasterClient)
{
    instance.chatRoom.OutputErrorNotMasterClient();
    return;
}

// Now we can assume that we are the master client
```

The remaining code has to be in one less if statement.

Lithrun commented on Assets/Scripts/Room/Chat/ChatCommands.cs:
```diff
@@ -0,0 +1,35 @@
+﻿public enum ChatCommand
```
Enums should start with a capital letter, following PascalCase

Lithrun reviewed (changes requested):
ChatCommandHandler is looking neat so far! A few code improvement tips in review 👍

Zoap committed:
resolving PR change requests

Zoap committed:
Merge branch 'development' into #166-Chat-Enhancements

Zoap committed:
.sln file

Zoap committed:
Merge from development

Zoap committed:
updated outputstrings for team switching , and removed some unnecessary comments

Zoap committed:
changes call from OnGui to Update

Co-authored-by: Wagner <mathiasvalentinwagner@gmail.com>

Zoap committed:
applying fix to PunTeams GetTeam()

Lithrun committed:
Temporary fix for chat input key

Zoap committed:
fixes joining individuals message having cyan color

Zoap committed:
Merge branch '#166-Chat-Enhancements' of https://github.com/AoTTG-2/AoTTG-2 into #166-Chat-Enhancements

Zoap committed:
- small refactoring
- removed the method for setting name color depending on team
- added global clearchat method

Lithrun committed:
Delete output_log.txt
